Alignment/MillePedeAlignmentAlgorithm/src/MillePedeFileReader.cc

```diff
@@ -298,10 +436,74 @@ std::string MillePedeFileReader::getStringFromHLS(MillePedeFileReader::PclHLS HL
+    case PclHLS::TPEPanelDiskM2:
+      return "TPEPanelDiskM2";
+    case PclHLS::TPEPanelDiskM3:
+      return "TPEPanelDiskM3";
     default:
+      //return "NotInPCL";
       throw cms::Exception("LogicError")
           << "@SUB=MillePedeFileReader::getStringFromHLS\n"
-          << "Found an alignable structure not possible to map in the default AlignPCLThresholds partitions";
+          << "Found an alignable structure not possible to map in the default AlignPCLThresholdsHG partitions";
+  }
+}
+
+int MillePedeFileReader::getIndexForHG(align::ID id, int detIndex) {
+  const auto& tns = pedeLabeler_->alignableTracker()->trackerNameSpace();
+  switch (detIndex) {
+    case 6:
+      return (tns.tpb().halfBarrelNumber(id) == 1) ? tns.tpb().ladderNumber(id) : tns.tpb().ladderNumber(id) + 6;
+    case 7:
+      return (tns.tpb().halfBarrelNumber(id) == 1) ? (tns.tpb().ladderNumber(id) + 12)
+                                                   : (tns.tpb().ladderNumber(id) + 12) + 14;
+    case 8:
+      return (tns.tpb().halfBarrelNumber(id) == 1) ? (tns.tpb().ladderNumber(id) + 40)
+                                                   : (tns.tpb().ladderNumber(id) + 40) + 22;
+    case 9:
+      return (tns.tpb().halfBarrelNumber(id) == 1) ? (tns.tpb().ladderNumber(id) + 84)
+                                                   : (tns.tpb().ladderNumber(id) + 84) + 32;
+    case 10:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 336
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 336 + 56;
+    case 11:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 448
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 448 + 56;
+    case 12:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 560
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 560 + 56;
+    case 13:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 224
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 224 + 56;
+    case 14:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 112
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 112 + 56;
+    case 15:
+      return (tns.tpe().halfCylinderNumber(id) == 1)
+                 ? (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148
+                 : (tns.tpe().bladeNumber(id) * 2 - (tns.tpe().panelNumber(id) % 2)) + 148 + 56;
```

Comment:
can we reduce here the amount of harcoded constants?

Reply:
I'm struggling a bit with this one. Probably it would be nice also to use the `PixelTopologyMap`, but if I'm not mistaken I can not take it from ES in the `MillePedeFileReader`, right?

Reply:
In principle you can make an object of the `PixelTopologyMap` class a data member of `MillePedeFileReader` that gets initialized elsewhere and then access here its methods, though if it slows you down too much, we can think on how to do it in a follow-up (of course if the other reviewers don't insist).

Reply:
I already thought about this, but the `MillePedeFileReader` is not only used in `MillePedeDQMModule` but also in `MillePedeAlignmentAlgorithm`, so there would be a few more changes when using the `PixelTopologyMap`. I am still not sure if it gets much better with this one.
So maybe we put this aside for now and come back to this later (hopefully with a better idea)

Reply:
> but also in MillePedeAlignmentAlgorithm

of course it would need to have all the callers changed, but presumably all of them have access to the ES.

Reply:
I came up with a solution, with almost no constants. Hope this is better now: https://github.com/cms-sw/cmssw/pull/38449/commits/6e9c819b448af26f5fef27a222238ff7c1f83c53